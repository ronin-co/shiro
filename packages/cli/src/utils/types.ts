import fs from 'node:fs/promises';

import json5 from 'json5';

/**
 * The name of the TypeScript declaration file stored inside the `.ronin` directory.
 */
export const TYPES_DTS_FILE_NAME = 'types.d.ts';

/**
 * The name of the Zod schemas file stored inside the `.ronin` directory.
 */
export const ZOD_SCHEMA_FILE_NAME = 'zod.ts';

/**
 * The name of the TypeScript declaration file stored inside the `.ronin` directory.
 */
const TYPES_INCLUDE_PATH = '.ronin/*.d.ts';

/**
 * Add the path to the generated TypeScript types to the `tsconfig.json` file.
 *
 * @param path - Path to the `tsconfig.json` file.
 *
 * @returns Promise resolving to void.
 */
export const injectTSConfigInclude = async (
  path: string,
): Promise<{
  compilerOptions: Record<string, unknown>;
  include: Array<string>;
}> => {
  // Set a base TypeScript config used for every project.
  const contents = {
    compilerOptions: {},
    include: new Array<string>(),
  };

  // Attempt to load the existing `tsconfig.json` file.
  const fileExists = await fs.exists(path);
  if (fileExists) {
    const fileContents = await fs.readFile(path, 'utf-8');
    const json = json5.parse(fileContents);
    Object.assign(contents, json);
  }

  // If the user has not already provided any `include` files then
  // we need to add some defaults so their code base continues to act
  // like it does currently.
  if (contents.include.length <= 0) contents.include.push('**/*.ts', '**/*.tsx');

  // Add the path to the generated TypeScript types to the `tsconfig.json` file.
  if (!contents.include.includes(TYPES_INCLUDE_PATH))
    contents.include.push(TYPES_INCLUDE_PATH);

  return contents;
};

/**
 * Generate the TypeScript types for a space.
 *
 * @param appTokenOrSessionToken - Authentication token used to authorize the API request.
 * @param slug - Slug of the space to generate types for.
 *
 * @returns Promise resolving to the generated TypeScript types.
 */
export const getSpaceTypes = async (
  appTokenOrSessionToken: string | undefined,
  slug: string,
): Promise<string> => {
  const url = new URL(`/generate/${slug}`, 'https://codegen.ronin.co/');
  url.searchParams.set('language', 'typescript');

  const response = await fetch(url.href, {
    headers: {
      Authorization: `Bearer ${appTokenOrSessionToken}`,
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) throw new Error(`API request failed with status: ${response.status}`);

  const code = await response.text();

  return code;
};

/**
 * Generate the Zod schemas for a space.
 *
 * @param appTokenOrSessionToken - Authentication token used to authorize the API request.
 * @param slug - Slug of the space to generate Zod schemas for.
 *
 * @returns Promise resolving to the generated Zod schemas.
 */
export const getZodSchemas = async (
  appTokenOrSessionToken: string | undefined,
  slug: string,
): Promise<string> => {
  const url = new URL(`/generate/${slug}`, 'https://codegen.ronin.co/');
  url.searchParams.set('language', 'zod');

  const response = await fetch(url.href, {
    headers: {
      Authorization: `Bearer ${appTokenOrSessionToken}`,
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) throw new Error(`API request failed with status: ${response.status}`);

  const code = await response.text();

  return code;
};
