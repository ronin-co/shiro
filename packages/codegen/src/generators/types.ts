import { SyntaxKind, addSyntheticLeadingComment, factory } from 'typescript';

import { genericIdentifiers, identifiers } from '@/src/constants/identifiers';
import { convertToPascalCase } from '@/src/utils/slug';
import { mapRoninFieldToTypeNode, remapNestedFields } from '@/src/utils/types';

import type {
  InterfaceDeclaration,
  TypeAliasDeclaration,
  TypeParameterDeclaration,
} from 'typescript';

import type { Model, ModelField } from '@/src/types/model';

const DEFAULT_FIELD_SLUGS = [
  'id',
  'ronin.createdAt',
  'ronin.createdBy',
  'ronin.locked',
  'ronin.updatedAt',
  'ronin.updatedBy',
] satisfies Array<string>;

/**
 * Generate all required type definitions for a provided RONIN model.
 *
 * This will generate a shared schema interface that is then used to create type
 * aliases for both the singular and plural model types.
 *
 * The plural model type will be mapped to an array of the singular model type
 * and extend it with the plural model properties.
 *
 * @param models - All RONIN models of the addressed space.
 *
 * @returns - An array of type nodes to be added to the `index.d.ts` file.
 */
export const generateTypes = (
  models: Array<Model>,
): Array<InterfaceDeclaration | TypeAliasDeclaration> => {
  const nodes = new Array<InterfaceDeclaration | TypeAliasDeclaration>();

  for (const model of models) {
    const fields: Array<ModelField> = Object.entries(model.fields)
      .map(([slug, field]) => ({ ...field, slug }) as ModelField)
      .filter((field) => !DEFAULT_FIELD_SLUGS.includes(field.slug));

    const singularModelIdentifier = factory.createIdentifier(
      convertToPascalCase(model.slug),
    );
    const pluralSchemaIdentifier = factory.createIdentifier(
      convertToPascalCase(model.pluralSlug),
    );

    const hasLinkFields = fields.some(
      (field) =>
        field.type === 'link' && models.some((model) => model.slug === field.target),
    );

    const modelInterfaceTypeParameters = new Array<TypeParameterDeclaration>();
    const linkFieldKeys = fields
      .filter((field) => field.type === 'link')
      .map((field) => {
        const literal = factory.createStringLiteral(field.slug);
        return factory.createLiteralTypeNode(literal);
      });

    /**
     * ```ts
     * <TUsing extends Array<...> | 'all' = []>
     * ```
     */
    const usingGenericDec = factory.createTypeParameterDeclaration(
      undefined,
      genericIdentifiers.using,
      factory.createUnionTypeNode([
        factory.createTypeReferenceNode(identifiers.primitive.array, [
          factory.createUnionTypeNode(linkFieldKeys),
        ]),
        factory.createLiteralTypeNode(factory.createStringLiteral('all')),
      ]),
      factory.createTupleTypeNode([]),
    );

    if (hasLinkFields) modelInterfaceTypeParameters.push(usingGenericDec);

    /**
     * ```ts
     * SchemaSlugSchema<TUsing>
     * ```
     */
    const modelSchemaName = factory.createTypeReferenceNode(
      singularModelIdentifier,
      hasLinkFields ? [factory.createTypeReferenceNode(genericIdentifiers.using)] : [],
    );

    /**
     * ```ts
     * export type SchemaSlug<TUsing extends Array<...> | 'all' = []> = ResultRecord & {
     *  // ...
     * };
     * ```
     */
    const singularModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      singularModelIdentifier,
      modelInterfaceTypeParameters,
      factory.createIntersectionTypeNode([
        factory.createExpressionWithTypeArguments(
          identifiers.syntax.resultRecord,
          undefined,
        ),
        factory.createTypeLiteralNode(
          remapNestedFields(fields)
            .sort(([a], [b]) => a.localeCompare(b))
            .map(([fieldSlug, field]) => {
              if (Array.isArray(field)) {
                const sortedFields = field.sort((a, b) => a.slug.localeCompare(b.slug));
                return factory.createPropertySignature(
                  undefined,
                  fieldSlug,
                  undefined,
                  factory.createTypeLiteralNode(
                    sortedFields.map((nestedField) =>
                      factory.createPropertySignature(
                        undefined,
                        nestedField.slug,
                        undefined,
                        factory.createUnionTypeNode(
                          mapRoninFieldToTypeNode(nestedField, models),
                        ),
                      ),
                    ),
                  ),
                );
              }

              return factory.createPropertySignature(
                undefined,
                fieldSlug,
                undefined,
                factory.createUnionTypeNode(mapRoninFieldToTypeNode(field, models)),
              );
            }),
        ),
      ]),
    );

    /**
     * ```ts
     * Array<SchemaSlug>;
     * ```
     */
    const pluralModelArrayTypeDec = factory.createTypeReferenceNode(
      identifiers.primitive.array,
      [modelSchemaName],
    );

    /**
     * ```ts
     * {
     *  moreBefore?: string;
     *  moreAfter?: string;
     * };
     * ```
     */
    const pluralModelPaginationPropsTypeDec = factory.createTypeLiteralNode([
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier('moreBefore'),
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier('moreAfter'),
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
    ]);

    /**
     * ```ts
     * export type SchemaPluralSlug<TUsing extends Array<...> | 'all' = []> = Array<SchemaSlug> & {
     *  moreBefore?: string;
     *  moreAfter?: string;
     * };
     * ```
     */
    const pluralModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      pluralSchemaIdentifier,
      modelInterfaceTypeParameters,
      factory.createIntersectionTypeNode([
        pluralModelArrayTypeDec,
        pluralModelPaginationPropsTypeDec,
      ]),
    );

    // If the model does not have a summary / description
    // then we can continue to the next iteration & not add any comments.
    if (!model.summary) {
      nodes.push(singularModelTypeDec, pluralModelTypeDec);
      continue;
    }

    nodes.push(
      addSyntheticLeadingComment(
        singularModelTypeDec,
        SyntaxKind.MultiLineCommentTrivia,
        `*\n * ${model.summary}\n `,
        true,
      ),
      addSyntheticLeadingComment(
        pluralModelTypeDec,
        SyntaxKind.MultiLineCommentTrivia,
        `*\n * ${model.summary}\n `,
        true,
      ),
    );
  }

  return nodes;
};
