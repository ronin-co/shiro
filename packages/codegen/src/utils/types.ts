import { SyntaxKind, factory } from 'typescript';

import { genericIdentifiers, identifiers } from '@/src/constants/identifiers';
import { MODEL_TYPE_TO_SYNTAX_KIND_KEYWORD } from '@/src/constants/schema';
import { convertToPascalCase } from '@/src/utils/slug';

import type { ModelField } from '@ronin/compiler';
import type { TypeNode } from 'typescript';

import type { Model } from '@/src/types/model';

/**
 * Map a RONIN model field to a TypeScript type node.
 *
 * @param field - The RONIN model field to map.
 * @param models - The list of all RONIN models. Used to resolve link fields.
 *
 * @returns An array of TypeScript type nodes representing the field type.
 */
export const mapRoninFieldToTypeNode = (
  field: ModelField,
  models: Array<Model>,
): Array<TypeNode> => {
  const propertyUnionTypes = new Array<TypeNode>();

  switch (field.type) {
    case 'link': {
      // Check to make sure the target model exists. If it doesn't we
      // fall back to using `unknown` as the type.
      const targetModel = models.find((model) => model.slug === field.target);
      if (!targetModel) {
        propertyUnionTypes.push(factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword));
        break;
      }

      // If the field is marked as `many` then we need to wrap the
      // type in an array.
      const schemaTypeRef = factory.createTypeReferenceNode(
        convertToPascalCase(targetModel.slug),
      );
      const resolvedLinkFieldNode = factory.createTypeReferenceNode(
        identifiers.utils.resolveSchema,
        [
          field.kind === 'many'
            ? factory.createTypeReferenceNode(identifiers.primitive.array, [
                schemaTypeRef,
              ])
            : schemaTypeRef,

          factory.createTypeReferenceNode(genericIdentifiers.using),

          factory.createLiteralTypeNode(factory.createStringLiteral(field.slug)),
        ],
      );

      propertyUnionTypes.push(resolvedLinkFieldNode);
      break;
    }
    case 'blob':
    case 'boolean':
    case 'date':
    case 'json':
    case 'number':
    case 'string': {
      const primitive = MODEL_TYPE_TO_SYNTAX_KIND_KEYWORD[field.type];
      propertyUnionTypes.push(primitive);
      break;
    }
    default: {
      propertyUnionTypes.push(factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword));
      break;
    }
  }

  // We need to mark fields as nullable if they are:
  // - Not required
  // - Not a link field
  // - Not a many-to-many link field
  if (field.required === false && field.type === 'link' && field.kind !== 'many')
    propertyUnionTypes.push(factory.createLiteralTypeNode(factory.createNull()));

  return propertyUnionTypes;
};

/**
 * Remap nested fields to a more usable format.
 *
 * Currently, nested fields for a RONIN model are stored using a dot-notation
 * slug, such as `foo.bar`. However, these are resolved as a nested object.
 * As such, we need to remap the fields to a more usable format.
 *
 * @param fields - The list of fields to remap.
 *
 * @returns An array of entries where the key is the field slug and the value is
 * either the field itself or an array of (nested) fields.
 */
export const remapNestedFields = (
  fields: Array<ModelField>,
): Array<[string, ModelField | Array<ModelField>]> => {
  const remappedFields = new Map<string, ModelField | Array<ModelField>>();

  for (const field of fields) {
    if (!field.slug.includes('.')) {
      remappedFields.set(field.slug, field);
      continue;
    }

    const [parentSlug, childSlug] = field.slug.split('.');
    const nestedField = Object.assign({}, field, {
      slug: childSlug,
    });

    const parentField = remappedFields.get(parentSlug);
    if (!parentField) {
      remappedFields.set(parentSlug, [nestedField]);
      continue;
    }

    if (!Array.isArray(parentField)) continue;

    parentField.push(nestedField);
  }

  return Array.from(remappedFields.entries());
};
