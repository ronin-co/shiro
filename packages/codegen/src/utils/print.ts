import {
  ListFormat,
  NewLineKind,
  ScriptKind,
  ScriptTarget,
  createPrinter,
  createSourceFile,
  factory,
} from 'typescript';

import { DEFAULT_FILE_NAME } from '@/src/constants/file';

import type { Node } from 'typescript';

interface PrintNodesOptions {
  /**
   * The name of the source file.
   *
   * @default 'index.d.ts'
   */
  fileName?: string;
  /**
   * Whether to remove comments from the output.
   *
   * @default false
   */
  removeComments?: boolean;
}

/**
 * Prints a list of nodes to a string.
 *
 * This is primarily used to actually output code pragmatically
 * generated to a string.
 *
 * @param nodes - The nodes to print.
 *
 * @returns The string representation of the nodes.
 */
export const printNodes = (nodes: Array<Node>, options?: PrintNodesOptions): string => {
  const { fileName = DEFAULT_FILE_NAME, removeComments = false } = options ?? {};

  // This is a mock file that is entirely in memory & blank. It is just
  // required for the printer to work. The printer will add the nodes to
  // this file & then return the string representation of the file.
  const sourceFile = createSourceFile(
    fileName,
    '',
    ScriptTarget.ESNext,
    true,
    ScriptKind.TS,
  );

  const printer = createPrinter({
    newLine: NewLineKind.LineFeed,
    removeComments,
  });

  const nodesArray = factory.createNodeArray(nodes);

  return printer.printList(ListFormat.MultiLine, nodesArray, sourceFile);
};
