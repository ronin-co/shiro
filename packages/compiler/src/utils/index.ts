import { handleBeforeOrAfter } from '@/src/instructions/before-after';
import { handleIncluding } from '@/src/instructions/including';
import { handleLimitedTo } from '@/src/instructions/limited-to';
import { handleOrderedBy } from '@/src/instructions/ordered-by';
import { handleSelecting } from '@/src/instructions/selecting';
import { handleTo } from '@/src/instructions/to';
import { handleUsing } from '@/src/instructions/using';
import { handleWith } from '@/src/instructions/with';
import {
  ROOT_MODEL_WITH_ATTRIBUTES,
  getModelBySlug,
  transformMetaQuery,
} from '@/src/model';
import type { InternalModelField, Model } from '@/src/types/model';
import type {
  CombinedInstructions,
  Instructions,
  InternalDependencyStatement,
  Query,
  SetInstructions,
  Statement,
} from '@/src/types/query';
import { DML_QUERY_TYPES_WRITE, QUERY_SYMBOLS } from '@/src/utils/constants';
import { RoninError, isObject, splitQuery } from '@/src/utils/helpers';
import { formatIdentifiers } from '@/src/utils/statement';

/**
 * Composes an SQL statement for a provided RONIN query.
 *
 * @param query - The RONIN query for which an SQL statement should be composed.
 * @param models - A list of models.
 * @param statementParams - A collection of values that will automatically be
 * inserted into the query by SQLite.
 * @param options - Additional options to adjust the behavior of the statement generation.
 *
 * @returns The composed SQL statement.
 */
export const compileQueryInput = (
  defaultQuery: Query,
  models: Array<Model>,
  // In order to prevent SQL injections and allow for faster query execution, we're not
  // inserting any values into the SQL statement directly. Instead, we will pass them to
  // SQLite's API later on, so that it can prepare an object that the database can
  // execute in a safe and fast manner. SQLite allows strings, numbers, and booleans to
  // be provided as values.
  statementParams: Array<unknown> | null,
  options?: {
    /** Whether the query should explicitly return records. Defaults to `true`. */
    returning?: boolean;
    /**
     * If the query is contained within another query, this option should be set to the
     * model of the parent query. Like that, it becomes possible to reference fields of
     * the parent model in the nested query (the current query).
     */
    parentModel?: Model;
    /**
     * Whether to compute default field values as part of the generated statement.
     */
    inlineDefaults: boolean;
  },
): {
  dependencies: Array<InternalDependencyStatement>;
  main: Statement;
  selectedFields: Array<InternalModelField>;
  model: Model;
  updatedQuery: Query;
} => {
  // A list of write statements that are required to be executed before the main read
  // statement. Their output is not relevant for the main statement, as they are merely
  // used to update the database in a way that is required for the main read statement
  // to return the expected results.
  const dependencyStatements: Array<InternalDependencyStatement> = [];

  // If the query is a meta query of type `create`, `alter`, or `drop`, we need to
  // transform it into a regular query before it can be processed further.
  const query = transformMetaQuery(
    models,
    dependencyStatements,
    statementParams,
    defaultQuery,
    {
      // biome-ignore lint/complexity/useSimplifiedLogicExpression: This is needed.
      inlineDefaults: options?.inlineDefaults || false,
    },
  );

  // If no further query processing should happen, we need to return early. This happens
  // when the root model is created.
  if (query === null)
    return {
      dependencies: [],
      main: dependencyStatements[0],
      selectedFields: [],
      model: ROOT_MODEL_WITH_ATTRIBUTES,
      updatedQuery: defaultQuery,
    };

  // Split out the individual components of the query.
  const parsedQuery = splitQuery(query);
  const { queryType, queryModel, queryInstructions } = parsedQuery;

  // Find the model that the query is interacting with.
  const model = getModelBySlug(models, queryModel);

  // Whether the query will interact with a single record, or multiple at the same time.
  const single = queryModel !== model.pluralSlug;

  // Walk deeper into the query, to the level on which the actual instructions (such as
  // `with` and `including`) are located.
  let instructions = formatIdentifiers(model, queryInstructions);

  const returning = options?.returning ?? true;

  // Apply any presets that are potentially being selected by the query.
  if (instructions && typeof instructions.using !== 'undefined') {
    instructions = handleUsing(model, instructions);
  }

  // If a `count` query was provided, ensure that the final select statement only includes
  // a single column that contains the amount of records that were matched by the query.
  if (queryType === 'count') {
    if (!instructions) instructions = {} as Instructions & SetInstructions;

    instructions!.selecting = ['amount'];
    instructions!.including = Object.assign(instructions?.including || {}, {
      amount: {
        [QUERY_SYMBOLS.EXPRESSION]: 'COUNT(*)',
      },
    });
  }

  // If a `limitedTo` instruction was provided, that means the amount of records returned
  // by the query will be limited to a specific amount, which, in turn, means that
  // pagination is activated automatically, so a cursor will be provided to the client
  // that can be used to retrieve the next page of records.
  //
  // Since `limitedTo` automatically activates pagination, we have to make sure that, if
  // the instruction is provided, we also automatically provide an `orderedBy`
  // instruction, as pagination requires the records to be ordered by at least one
  // specific column, otherwise the cursor wouldn't work, since the order of the rows
  // might differ between pages.
  if (
    !single &&
    ((queryType === 'get' && instructions?.limitedTo) ||
      (queryType === 'count' && (instructions?.before || instructions?.after)))
  ) {
    instructions = instructions || {};
    instructions.orderedBy = instructions.orderedBy || {};
    instructions.orderedBy.ascending = instructions.orderedBy.ascending || [];
    instructions.orderedBy.descending = instructions.orderedBy.descending || [];

    if (
      ![
        ...instructions.orderedBy.ascending,
        ...instructions.orderedBy.descending,
      ].includes('ronin.createdAt')
    ) {
      // It's extremely important that the item is added to the end of the array,
      // otherwise https://linear.app/ronin/issue/RON-1084 would occur.
      instructions.orderedBy.descending.push('ronin.createdAt');
    }
  }

  // A list of columns that should be selected when querying records.
  const { columns, isJoining, selectedFields } = handleSelecting(
    models,
    model,
    statementParams,
    queryType,
    single,
    {
      selecting: instructions?.selecting,
      including: instructions?.including,
      orderedBy: instructions?.orderedBy,
      limitedTo: instructions?.limitedTo,
    },
    // biome-ignore lint/complexity/useSimplifiedLogicExpression: This is needed.
    { inlineDefaults: options?.inlineDefaults || false },
  );

  let statement = '';

  switch (queryType) {
    case 'get':
    case 'count':
      statement += `SELECT ${columns} FROM `;
      break;

    case 'set':
      statement += 'UPDATE ';
      break;

    case 'add':
      statement += 'INSERT INTO ';
      break;

    case 'remove':
      statement += 'DELETE FROM ';
      break;
  }

  let isJoiningMultipleRows = false;

  if (isJoining) {
    const { statement: including, tableSubQuery } = handleIncluding(
      models,
      model,
      statementParams,
      single,
      instructions?.including,
    );

    // If multiple rows are being joined from a different table, even though the root
    // query is only supposed to return a single row, we need to ensure a limit for the
    // root query *before* joining the other rows. Otherwise, if the limit sits at the
    // end of the full query, only one row would be available at the end.
    if (tableSubQuery) {
      statement += `(${tableSubQuery}) as ${model.tableAlias} `;
      isJoiningMultipleRows = true;
    } else {
      statement += `"${model.table}" `;
    }

    statement += `${including} `;
  } else {
    statement += `"${model.table}" `;
  }

  if (queryType === 'add' || queryType === 'set') {
    const instructionName = queryType === 'add' ? 'with' : 'to';
    const instructionValue = instructions![instructionName] as CombinedInstructions['to'];

    // This validation must be performed before any default fields (such as `ronin`) are
    // added to the record. Otherwise there are always fields present.
    if (
      !(instructionValue && isObject(instructionValue)) ||
      Object.keys(instructionValue).length === 0
    ) {
      throw new RoninError({
        message: `When using a \`${queryType}\` query, the \`${instructionName}\` instruction must be a non-empty object.`,
        code: instructionName === 'to' ? 'INVALID_TO_VALUE' : 'INVALID_WITH_VALUE',
        queries: [query],
      });
    }

    const toStatement = handleTo(
      models,
      model,
      statementParams,
      queryType,
      dependencyStatements,
      { with: instructions!.with, to: instructionValue },
      options,
    );

    statement += `${toStatement} `;
  }

  const conditions: Array<string> = [];

  // Queries of type "get", "set", "remove", or "count" all support filtering records,
  // but those of type "add" do not.
  if (queryType !== 'add' && instructions && Object.hasOwn(instructions, 'with')) {
    const withStatement = handleWith(
      models,
      model,
      statementParams,
      instructions!.with,
      options?.parentModel,
    );

    if (withStatement.length > 0) conditions.push(withStatement);
  }

  if (
    instructions &&
    (typeof instructions.before !== 'undefined' ||
      typeof instructions.after !== 'undefined')
  ) {
    if (single) {
      throw new RoninError({
        message:
          'The `before` and `after` instructions are not supported when querying for a single record.',
        code: 'INVALID_BEFORE_OR_AFTER_INSTRUCTION',
        queries: [query],
      });
    }

    const beforeAndAfterStatement = handleBeforeOrAfter(
      model,
      statementParams,
      queryType,
      {
        before: instructions.before,
        after: instructions.after,
        with: instructions.with,
        orderedBy: instructions.orderedBy,
        limitedTo: instructions.limitedTo,
      },
    );

    conditions.push(beforeAndAfterStatement);
  }

  if (conditions.length > 0) {
    // If multiple conditions are available, wrap them in parentheses to ensure that the
    // AND/OR comparisons are asserted correctly.
    if (conditions.length === 1) {
      statement += `WHERE ${conditions[0]} `;
    } else {
      statement += `WHERE (${conditions.join(' ')}) `;
    }
  }

  if (instructions?.orderedBy) {
    const orderedByStatement = handleOrderedBy(model, instructions.orderedBy);
    statement += `${orderedByStatement} `;
  }

  if (
    queryType === 'get' &&
    !isJoiningMultipleRows &&
    (single || instructions?.limitedTo)
  ) {
    statement += handleLimitedTo(single, instructions?.limitedTo);
  }

  // For queries that modify records, we want to make sure that the modified record is
  // returned after the modification has been performed.
  if ((DML_QUERY_TYPES_WRITE as ReadonlyArray<string>).includes(queryType) && returning) {
    statement += `RETURNING ${columns}`;
  }

  const mainStatement: Statement = {
    statement: statement.trimEnd(),
    params: statementParams || [],
  };

  // We are setting this property separately to make sure it doesn't even exist if the
  // query doesn't return any output. This makes it easier for developers to visually
  // distinguish queries that return output from those that don't, when looking at the
  // output produced by the compiler.
  if (returning) mainStatement.returning = true;

  return {
    dependencies: dependencyStatements,
    main: mainStatement,
    selectedFields,
    model,
    updatedQuery: query,
  };
};
