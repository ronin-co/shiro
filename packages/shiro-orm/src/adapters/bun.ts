import { Database } from 'bun:sqlite';

import type { Model } from 'shiro-compiler';

interface ShiroOptions<T extends Array<Model>> {
  database?: Database;
  models: T;
}

/**
 * @todo Add documentation
 */
export const shiro = <T extends Array<Model>>(options: ShiroOptions<T>) => {
  const database = options.database ?? new Database(':memory:');
};
