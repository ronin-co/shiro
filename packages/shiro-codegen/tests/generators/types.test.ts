import { describe, expect, test } from 'bun:test';
import {
  blob,
  boolean,
  date,
  json,
  link,
  model,
  number,
  string,
} from 'shiro-syntax/schema';

import { generateTypes } from '@/src/generators/types';
import { printNodes } from '@/src/utils/print';

describe('types', () => {
  test('a basic model', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        avatar: blob(),
        email: string({ required: true }),
        isActive: boolean(),
        lastActiveAt: date(),
        name: string(),
        rewardPoints: number({ defaultValue: 0, required: true }),
        settings: json({ defaultValue: {}, required: true }),
      },
    });

    // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
    // @ts-expect-error Codegen models types differ from the schema model types.
    const typesResult = generateTypes([AccountModel], AccountModel);

    expect(typesResult).toHaveLength(2);

    const typesResultStr = printNodes(typesResult);

    expect(typesResultStr).toMatchSnapshot();
  });

  test('a model with a summary', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: string(),
        email: string({ required: true }),
      },
      // @ts-expect-error This property is not native to RONIN models.
      summary: 'A user account.',
    });

    // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
    // @ts-expect-error Codegen models types differ from the schema model types.
    const typesResult = generateTypes([AccountModel], AccountModel);

    expect(typesResult).toHaveLength(2);

    const typesResultStr = printNodes(typesResult);

    expect(typesResultStr).toMatchSnapshot();
  });

  test('a model with an invalid field type', () => {
    const field = string();

    // @ts-expect-error These properties are designed to be read-only.
    field.type = 'invalid';

    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: field,
      },
    });

    // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
    // @ts-expect-error Codegen models types differ from the schema model types.
    const typesResult = generateTypes([AccountModel], AccountModel);

    expect(typesResult).toHaveLength(2);

    const typesResultStr = printNodes(typesResult);

    expect(typesResultStr).toMatchSnapshot();
  });

  test('a model with a link field', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: string(),
        email: string({ required: true }),
      },
    });

    const PostModel = model({
      slug: 'post',
      pluralSlug: 'posts',
      fields: {
        title: string(),
        author: link({ target: 'account' }),
      },
    });

    // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
    // @ts-expect-error Codegen models types differ from the schema model types.
    const typesResult = generateTypes([AccountModel, PostModel], PostModel);

    expect(typesResult).toHaveLength(4);

    const typesResultStr = printNodes(typesResult);

    expect(typesResultStr).toMatchSnapshot();
  });

  test('a model with a many-to-many link field', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: string(),
        email: string({ required: true }),
      },
    });

    const SpaceModel = model({
      slug: 'space',
      pluralSlug: 'spaces',
      fields: {
        name: string(),
        members: link({ target: 'account', kind: 'many' }),
      },
    });

    // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
    // @ts-expect-error Codegen models types differ from the schema model types.
    const typesResult = generateTypes([AccountModel, SpaceModel], SpaceModel);

    expect(typesResult).toHaveLength(4);

    const typesResultStr = printNodes(typesResult);

    expect(typesResultStr).toMatchSnapshot();
  });

  test('a model with a link field that does not exist', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: string(),
        email: string({ required: true }),
        latestPost: link({ target: 'does_not_exist' }),
      },
    });

    // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
    // @ts-expect-error Codegen models types differ from the schema model types.
    const typesResult = generateTypes([AccountModel], AccountModel);

    expect(typesResult).toHaveLength(2);

    const typesResultStr = printNodes(typesResult);

    expect(typesResultStr).toMatchSnapshot();
  });

  test('a model with nested fields', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: string(),
        'nested.foo': string(),
        'nested.bar': number(),
      },
      // @ts-expect-error This property is not native to RONIN models.
      summary: 'A user account.',
    });

    // TODO(@nurodev): Refactor the `Model` type to be more based on current schema models.
    // @ts-expect-error Codegen models types differ from the schema model types.
    const typesResult = generateTypes([AccountModel], AccountModel);

    expect(typesResult).toHaveLength(2);

    const typesResultStr = printNodes(typesResult);

    expect(typesResultStr).toMatchSnapshot();
  });
});
