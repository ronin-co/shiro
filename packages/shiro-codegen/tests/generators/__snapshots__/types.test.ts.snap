// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`types a basic model 1`] = `
"export type Account = ResultRecord & {
    avatar: StoredObject;
    email: string;
    isActive: boolean;
    lastActiveAt: Date;
    name: string;
    rewardPoints: number;
    settings: (JsonObject | JsonArray);
};
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types a model with a summary 1`] = `
"/**
 * A user account.
 */
export type Account = ResultRecord & {
    email: string;
    name: string;
};
/**
 * A user account.
 */
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types a model with an invalid field type 1`] = `
"export type Account = ResultRecord & {
    name: unknown;
};
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types a model with a link field 1`] = `
"export type Account = ResultRecord & {
    email: string;
    name: string;
};
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
export type Post<TUsing extends Array<"author"> | "all" = [
]> = ResultRecord & {
    author: ResolveSchema<Account, TUsing, "author">;
    title: string;
};
export type Posts<TUsing extends Array<"author"> | "all" = [
]> = Array<Post<TUsing>> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types a model with a many-to-many link field 1`] = `
"export type Account = ResultRecord & {
    email: string;
    name: string;
};
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
export type Space<TUsing extends Array<"members"> | "all" = [
]> = ResultRecord & {
    members: ResolveSchema<Array<Account>, TUsing, "members">;
    name: string;
};
export type Spaces<TUsing extends Array<"members"> | "all" = [
]> = Array<Space<TUsing>> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types a model with a link field that does not exist 1`] = `
"export type Account = ResultRecord & {
    email: string;
    latestPost: unknown;
    name: string;
};
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;

exports[`types a model with nested fields 1`] = `
"/**
 * A user account.
 */
export type Account = ResultRecord & {
    name: string;
    nested: {
        bar: number;
        foo: string;
    };
};
/**
 * A user account.
 */
export type Accounts = Array<Account> & {
    moreBefore?: string;
    moreAfter?: string;
};
"
`;
